BlueprintUI/Sources/Element/ElementState.swift

```diff
@@ -0,0 +1,597 @@
+        if let existing = measurements[constraint] {
+            return existing.size
+        }
+
+        /// Perform the measurement and track the environment dependencies.
+        let (size, dependencies) = trackEnvironmentReads(for: .measurement, with: environment, in: measurer)
+
+        /// Save the measurement for next time.
+        measurements[constraint] = .init(
+            size: size,
+            dependencies: dependencies
+        )
+
+        return size
+    }
+
+    /// Represents a cached `LayoutResultNode` tree of all children,
+    /// which contains both child tree, and the dependencies
+    /// from the `Environment`, if any.
+    private struct CachedLayout {
+        var layout: [LayoutResultNode]
+        var dependencies: Environment.Subset?
+    }
+
+    /// Invoked by `ElementContent` to generate a layout tree via the `layout`,
+    /// or return a cached layout tree if one exists.
+    func layout(
+        in size: CGSize,
+        with environment: Environment,
+        using layout: (Environment) -> [LayoutResultNode]
+    ) -> [LayoutResultNode] {
+
+        /// Layout is only invoked once per cycle. If we're not a `ComparableElement`,
+        /// there's no point in caching this value, so just return the `layout` directly.
+        guard comparability == .comparableElement else {
+            return layout(environment)
+        }
+
+        /// We already have a cached layout, reuse that.
+        if let existing = layouts[size] {
+            return existing.layout
+        }
+
+        /// Perform the layout and track the environment dependencies.
+        let (layout, dependencies) = trackEnvironmentReads(for: .layout, with: environment, in: layout)
+
+        /// Save the layout for next time.
+        layouts[size] = .init(
+            layout: layout,
+            dependencies: dependencies
+        )
+
+        return layout
+    }
+
+    /// Used by measurement and layout to track the dependencies they have on the `Environment`.
+    /// The method returns the output from your `toTrack` callback, as well as
+    /// an `Environment.Subset` that represents the measurement or layout's dependencies on the `Environment`.
+    ///
+    /// If there are no dependencies (`toTrack` did not read from the `Environment`), no subset is returned.
+    ///
+    /// Additionally, if the element is not comparable, no subset is returned or tracked.
+    private func trackEnvironmentReads<Output>(
+        for layoutPass: Environment.LayoutPass,
+        with environment: Environment,
+        in toTrack: (Environment) -> Output
+    ) -> (Output, Environment.Subset?) {
+
+        switch layoutPass {
+        case .measurement:
+            if comparability.isComparable {
+                break
+            } else {
+                return (toTrack(environment), nil)
+            }
+        case .layout:
+            if comparability == .comparableElement {
```

Comment:
Same here about noting why `.childOfComparableElement` is treated differently from `.comparableElement` (or just move this logic into the call site so we can document this behavior and reason about it in one place).

Reply:
Done